x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/attribute_selector.tsx

```diff
@@ -85,7 +85,7 @@ const getSelectedOptions = (
   const options: Array<EuiComboBoxOptionOption<string>> = groupedOptions.reduce(
     (acc: Array<EuiComboBoxOptionOption<string>>, n: EuiComboBoxOptionOption<string>) => [
       ...acc,
-      ...(n.options || []),
+      ...(n.options as Array<EuiComboBoxOptionOption<string>>),
```

Comment:
I don't remember adding this originally, to be honest.

I think casting like this to get around a TS error is usually a code smell. The type is like "Hey, I could be undefined, better make sure you handle me!". And by casting, you're basically like "I call BS, you'll never be undefined".

Which is fine sometimes, but I think you just need to make extra sure it won't actually ever be undefined.

Reply:
One thing you *could* do that might clean this up is to declare you own type here that extends EuiComboBoxOptionOption, that doesn't have optional value and options:

```
interface EuiComboBoxOptionOption<T> {
    value?: T;
    options?: Array<{ foo: string }>
}

interface Option extends EuiComboBoxOptionOption<string> {
    value: string;
    options: Array<{ foo: string }>
}

let a: EuiComboBoxOptionOption<string> = {
    value: 'a',
    options: [{foo: 'foo'}]
}

let b: Option = {
    value: 'a',
    options: [{foo: 'foo'}]
}

// Can't do this because a might be undefined
// a.value.toUpperCase();

// Now you don't have to check if value is defined, it's always defined because you're using the "Option" type
b.value.toUpperCase();
```

Reply:
Actually, I think you'd want types like this:
```
interface ParentOption extends EuiComboBoxOptionOption<string> {
  label: string;
  options: ChildOption[];
}

interface ChildOption extends EuiComboBoxOptionOption<string> {
  value: string;
  label: string;
}
```

Reply:
Which would let you simplify that method to:

```
const getSelectedOptions = (selectedAuthProviders: string[], availableAuthProviders: string[]) => {
  const groupedOptions: ParentOption[] = getAuthProviderOptions(availableAuthProviders);
  const childOptions: ChildOption[] = [];
  const options = groupedOptions.reduce((acc, n) => [...acc, ...n.options], childOptions);
  return options.filter((o) => o.value && selectedAuthProviders.includes(o.value));
};
```

Reply:
And your second cast would be:
```
onChange={(options) => {
  handleAuthProviderChange(options.map((o) => (o as ChildOption).value));
}}
```

Reply:
I like that better, seems a bit cleaner. Let me know what you think. I have it locally and it looks 👌 .

Reply:
Thanks Jason! I agree it's a lot cleaner. I am seeing this locally when I implement these changes:
![image](https://user-images.githubusercontent.com/1869731/108429455-9e3af900-7205-11eb-8bd9-e8cc125e2aff.png)
You're not seeing that locally?

Reply:
One of the issues I had was problems trying to reproduce this in the browser locally. I had never set up multiple auths before and in the process of figuring out how, it occurred to me that we are going to be using Kibana exclusively for authentication post 8.x. Do we even need this code here anymore? I'm a bit hazy on how auth works in general and how that will translate to when we are no longer logging in via the legacy app.

Reply:
Here's my full diff

```diff
diff --git a/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/attribute_selector.tsx b/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/attribute_selector.tsx
index 301815edf28..94ff78e3aa9 100644
--- a/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/attribute_selector.tsx
+++ b/x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/attribute_selector.tsx
@@ -20,6 +20,16 @@ import {
   EuiTitle,
 } from '@elastic/eui';
 
+interface ParentOption extends EuiComboBoxOptionOption<string> {
+  label: string;
+  options: ChildOption[];
+}
+
+interface ChildOption extends EuiComboBoxOptionOption<string> {
+  value: string;
+  label: string;
+}
+
 import {
   ANY_AUTH_PROVIDER,
   ANY_AUTH_PROVIDER_OPTION_LABEL,
@@ -57,9 +67,7 @@ const attributeValueExamples = {
   metadata: '{"_reserved": true}',
 };
 
-const getAuthProviderOptions = (
-  availableAuthProviders: string[]
-): Array<EuiComboBoxOptionOption<string>> => {
+const getAuthProviderOptions = (availableAuthProviders: string[]) => {
   return [
     {
       label: AUTH_ANY_PROVIDER_LABEL,
@@ -75,20 +83,10 @@ const getAuthProviderOptions = (
   ];
 };
 
-const getSelectedOptions = (
-  selectedAuthProviders: string[],
-  availableAuthProviders: string[]
-): Array<EuiComboBoxOptionOption<string>> => {
-  const groupedOptions: Array<EuiComboBoxOptionOption<string>> = getAuthProviderOptions(
-    availableAuthProviders
-  );
-  const options: Array<EuiComboBoxOptionOption<string>> = groupedOptions.reduce(
-    (acc: Array<EuiComboBoxOptionOption<string>>, n: EuiComboBoxOptionOption<string>) => [
-      ...acc,
-      ...(n.options as Array<EuiComboBoxOptionOption<string>>),
-    ],
-    []
-  );
+const getSelectedOptions = (selectedAuthProviders: string[], availableAuthProviders: string[]) => {
+  const groupedOptions: ParentOption[] = getAuthProviderOptions(availableAuthProviders);
+  const childOptions: ChildOption[] = [];
+  const options = groupedOptions.reduce((acc, n) => [...acc, ...n.options], childOptions);
   return options.filter((o) => o.value && selectedAuthProviders.includes(o.value));
 };
 
@@ -124,7 +122,7 @@ export const AttributeSelector: React.FC<IAttributeSelectorProps> = ({
                 selectedOptions={getSelectedOptions(selectedAuthProviders, availableAuthProviders)}
                 options={getAuthProviderOptions(availableAuthProviders)}
                 onChange={(options) => {
-                  handleAuthProviderChange(options.map((o) => o.value as string));
+                  handleAuthProviderChange(options.map((o) => (o as ChildOption).value));
                 }}
                 fullWidth
                 isDisabled={disabled}
```

Regarding yours second comment, it my be best if we talk through this on Zoom. My memory is a bit hazy, and I'd like to see how that code actually translate to UI.

Reply:
Thanks Jason. The issue was the return type on `getAuthProviderOptions`. Removing that to match your diff fixed the error. 

In regards to the second comment, As long as it works in Kibana, I am fine leaving it, at least until we make final decisions on auth and once we are out of beta, since our aim was parity in 8.0 an we might have some super edge cases that need this in the beta, while we are supporting both. 

Thanks for your time! Would you mind reviewing on behalf of App Search?

Reply:
Can do.